### PR TITLE
fix(auto-tile): Add M/N tiling for loop-carried matmul_acc

### LINKED ISSUE #2232: [Pass Bug] AutoTileMatmulL0: matmul_acc M/N deferral drops the supported K tiling, then fails with a misleading buffer error
## Background

In pypto-lib, `models/deepseek/v4-pro/qkv_proj_rope.py` on `a2a3` has a **compile** problem: the `qproj` cube scope cannot use any N tile wider than 1024, and every wider tile fails with a buffer error that names a buffer the kernel author never sized (L0b / L1, or a ptoas fractal-alignment error) rather than the L0c accumulator that actually caused it.

This surfaced while tuning the `qproj` matmul tile: the kernel accumulates over K with the idiomatic split-K form (`pl.matmul` for the first K block, `pl.matmul_acc` for the rest), and every attempt to widen `QPROJ_MM_N_TILE` past 1024 failed with an unrelated-looking error.

The reproduction case below depends on **pypto only** — no torch, no device, no pypto-lib.

Reproduce with:

    python repro_pypto_only.py <N_TILE> [M_TILE]

The repro case is **not committed** — save the file below and run it directly. Files required: `repro_pypto_only.py`.

<details>
<summary>Reproduction case — <code>repro_pypto_only.py</code> (save anywhere; pypto-only)</summary>

```python
"""Minimal reproducer: AutoTileMatmulL0 defers M/N tiling for `tile.matmul_acc`
(PH-AT-006) and leaves the matmul ENTIRELY untiled -- discarding the K tiling it
would otherwise apply -- so a downstream buffer overflows and the reported error
names a buffer the kernel author never sized.

Depends on pypto only (no torch, no device, no pypto-lib).

    python repro_pypto_only.py <N_TILE> [M_TILE]     # default M_TILE=16
"""

import sys

import pypto.language as pl
from pypto import backend as _backend
from pypto.backend import BackendType

_backend.set_backend_type(BackendType.Ascend910B)

N_TILE = int(sys.argv[1]) if len(sys.argv) > 1 else 1152
M_TILE = int(sys.argv[2]) if len(sys.argv) > 2 else 16
K_TOTAL = 1024
K_TILE = 128
N_TOTAL = N_TILE * 8


@pl.jit
def repro(
    a: pl.Tensor[[M_TILE, K_TOTAL], pl.INT8],
    b: pl.Tensor[[K_TOTAL, N_TOTAL], pl.INT8],
    c: pl.Out[pl.Tensor[[M_TILE, N_TOTAL], pl.INT32]],
):
    for i in pl.spmd(N_TOTAL // N_TILE, name_hint="mm"):
        n0 = i * N_TILE
        acc = pl.create_tensor([M_TILE, N_TILE], dtype=pl.INT32)
        for kb in pl.pipeline(0, K_TOTAL // K_TILE, stage=2):
            k0 = kb * K_TILE
            at = a[0:M_TILE, k0 : k0 + K_TILE]
            bt = b[k0 : k0 + K_TILE, n0 : n0 + N_TILE]
            if k0 == 0:
                acc = pl.matmul(at, bt, out_dtype=pl.INT32)
            else:
                acc = pl.matmul_acc(acc, at, bt)
        c[0:M_TILE, n0 : n0 + N_TILE] = acc
    return c


if __name__ == "__main__":
    acc_bytes = ((M_TILE + 31) // 32) * 32 * N_TILE * 4
    label = (f"M={M_TILE:<3} N={N_TILE:<5} AlignUp(M,32)*N*4={acc_bytes:>7}B (L0c=131072B)  "
             f"L0B[{K_TILE},{N_TILE}]i8={K_TILE * N_TILE:>7}B (L0b=65536B)")
    try:
        repro.compile()
        print(f"{label} -> COMPILED")
    except Exception as e:  # noqa: BLE001 - the point of the repro
        msg = str(e).replace("\n", " ")
        for pat in ("Right buffer usage", "Acc buffer usage", "Mat buffer usage",
                    "Left buffer usage", "op expects"):
            i = msg.find(pat)
            if i >= 0:
                print(f"{label} -> FAILED: {msg[i:i + 100]}")
                break
        else:
            print(f"{label} -> FAILED: {type(e).__name__}: {msg[:120]}")
```

</details>

Reproduction environment:

| Component | Version |
|---|---|
| pypto-lib | `0442b08` (branch: `main`) |
| pypto | `f27c105a` (branch: `main`) |
| simpler | `9922afdb` (branch: detached) |
| ptoas | `0.54` |
| pto-isa | `83d01313` |
| CANN | `8.3.T14.0.B101:8.3.RC1.alpha003` |

All components match their pins (simpler gitlink, `pto_isa.pin`, `PTOAS_VERSION`).

Diagnosis: **pypto** — the deferral decision and the all-or-nothing fallback are both in `AutoTileMatmulL0`.

### Pass Name

AutoTileMatmulL0

### Description

`AutoTileMatmulL0` cannot M/N-tile `tile.matmul_acc` (a documented deferral, `PH-AT-006`, "needs per-sub-tile accumulator slicing"). That limitation is understood — the bug is the **fallback behaviour** and the **diagnostic** when it triggers:

1. **The deferral is all-or-nothing.** When the `[M, N]` accumulator exceeds L0c the pass needs M/N tiling, sees `matmul_acc`, and leaves the matmul *entirely untouched* — discarding the **K tiling it would otherwise have applied and which is fully supported for `matmul_acc`**. The untiled `[K_tile, N_tile]` B operand then reaches `AllocateMemoryAddr` and overflows L0b.

   This is demonstrably a lost opportunity, not an impossibility: at `N_TILE=1024` the pass *does* K-tile this exact kernel (choosing `k=32`, visible as "32768 B per stage" in the `PH-MR-001` hint). At `N_TILE=1152` the same `k=32` split would give `32 * 1152 = 36864 B <= 65536 B` and fit L0b comfortably — but the M/N question aborts the whole rewrite before the K tiling is ever attempted.

2. **The diagnostic names the wrong buffer.** One root cause produces five different messages naming three different buffers, and none of them mentions the L0c accumulator that forced the deferral. The `PH-AT-006` hint that *does* explain it is written to `report/perf_hints.log`, which is not surfaced on the failure path.

The effective ceiling is `AlignUp(M, 32) * N * 4 <= L0c`, i.e. **N <= 1024** for an INT32 accumulator on 910B, independent of M below 32. `M_TILE=16` and `M_TILE=32` produce byte-identical boundaries, so a 16-row accumulator is charged as a padded 32-row one (consistent with the `innerCols (32)` in the ptoas message).

For a decode-shaped kernel (few real rows, wide N) this caps the cube N-fragment at 1024 and forces a hand-written workaround: shrinking the DSL K tile to 32 so the *untiled* fragment happens to fit L0b — i.e. doing manually the tiling the pass declined.

### Git Commit ID

f27c105a0791d9038a9b8da48e4df7590f351f23

### Before IR (Input)

The reproducer above, in `@pl.jit` form (the shape the bug is actually hit in). The relevant region is one Mat-resident split-K accumulation whose `[M, N]` INT32 accumulator is sized by `N_TILE`:

```python
for i in pl.spmd(N_TOTAL // N_TILE, name_hint="mm"):
    n0 = i * N_TILE
    acc = pl.create_tensor([M_TILE, N_TILE], dtype=pl.INT32)
    for kb in pl.pipeline(0, K_TOTAL // K_TILE, stage=2):
        k0 = kb * K_TILE
        at = a[0:M_TILE, k0 : k0 + K_TILE]
        bt = b[k0 : k0 + K_TILE, n0 : n0 + N_TILE]
        if k0 == 0:
            acc = pl.matmul(at, bt, out_dtype=pl.INT32)
        else:
            acc = pl.matmul_acc(acc, at, bt)      # <-- M/N tiling deferred here
    c[0:M_TILE, n0 : n0 + N_TILE] = acc
```

### Expected IR (After Transformation)

Either of the following would be acceptable; (a) is the real fix, (b) is a cheap mitigation.

**(a) Implement M/N tiling for `tile.matmul_acc`** (per-sub-tile slicing of the caller's `[M, N]` accumulator), so wide-N tiles lower like the plain-`tile.matmul` path already does and the N <= 1024 ceiling disappears.

**(b) Degrade gracefully — keep the K tiling when M/N tiling is unavailable.** For `N_TILE=1152` the pass should emit the same K-tiled rewrite it already emits at `N_TILE=1024` — a `ForKind::Pipeline` K-loop over `tile.extract(..., target_memory=Right)` slices with `k=32`:

```
L0b per stage = k * N_TILE = 32 * 1152 = 36864 B  <=  65536 B      # fits
```

i.e. the untiled `[128, 1152]` extract is replaced by four `[32, 1152]` extracts, exactly as at `N_TILE=1024`. The kernel then compiles instead of failing.

If neither is possible for a given shape, the pass should **fail at the point of deferral** with a message naming the real constraint — the L0c accumulator, its shape, and the implied N cap — rather than letting a downstream allocator blame L0b/L1.

### Actual IR or Error

One root cause, five messages, three different buffers. `K_TILE=128`, INT8 operands, INT32 accumulator, Ascend910B (L0c 128 KB, L0a/L0b 64 KB):

```text
M=16  N=1024  AlignUp(M,32)*N*4= 131072B (L0c=131072B) -> COMPILED
M=32  N=1024  AlignUp(M,32)*N*4= 131072B (L0c=131072B) -> COMPILED

M=16  N=1152  AlignUp(M,32)*N*4= 147456B (L0c=131072B) -> FAILED
  ptoas compilation failed: error: 'pto.alloc_tile' op expects result boxed tile
  cols to be a multiple of innerCols (32), but got 16
  Error: Failed to parse MLIR.

M=32  N=1152  AlignUp(M,32)*N*4= 147456B (L0c=131072B) -> FAILED
  Right buffer usage (147456 bytes) exceeds platform limit (65536 bytes)

M=16  N=1408  AlignUp(M,32)*N*4= 180224B (L0c=131072B) -> FAILED
  Right buffer usage (180224 bytes) exceeds platform limit (65536 bytes)

M=16  N=2048  AlignUp(M,32)*N*4= 262144B (L0c=131072B) -> FAILED
  Mat buffer usage (528384 bytes) exceeds platform limit (524288 bytes)

M=32  N=2048  AlignUp(M,32)*N*4= 262144B (L0c=131072B) -> FAILED
  InternalError: tensor.create: tile size (262144 bytes) exceeds buffer capacity
  (188416 bytes) for memory space 1
```

The only place the real cause appears is `build_output/<case>/report/perf_hints.log`, which is not shown on the failure path:

```text
[perf_hint PH-AT-006] AutoTileMatmulL0: tile.matmul_acc with an oversized [M, N]
output needs M/N tiling -- the matmul_acc path is deferred (needs per-sub-tile
accumulator slicing); left untouched at <repro>:47:4
```

For contrast, at `N_TILE=1024` (compiling case) there is **no** `PH-AT-006` — only `PH-MR-001`, whose "32768 B per stage" confirms the pass chose `k=32` and K-tiled successfully:

```text
[perf_hint PH-MR-001] MemoryReuse: software pipelining requested depth 2 for
pipeline group 0 in Right, but only 1 of 2 buffers fit (32768 B per stage,
65536 B free) ...
```

### NPU Kind (if relevant)

Ascend 910B

### Host Platform (if relevant)

Linux (aarch64)

### Additional Context

Relevant sources:

- `src/ir/transforms/auto_tile_matmul_l0_pass.cpp:91-96` — documents M/N tiling as supported for `tile.matmul` but deferred for `tile.matmul_acc`.
- `src/ir/transforms/auto_tile_matmul_l0_pass.cpp:1187,1315` — the `PH-AT-006` emit sites.
- `docs/en/dev/passes/15-auto_tile_matmul_l0.md:55` and the `PH-AT-006` row at `:252` — same statement in the pass docs, including "left untouched".
- `include/pypto/backend/910B/backend_910b_handler.h:67` — `GetL0cCapacityBytes() == 128 KB`.

Impact: on a decode-shaped DeepSeek-V4 `qproj` (`M=16` after row padding, `N = H*HEAD_DIM`, INT8 W8A8), this caps the cube N-fragment at 1024 and made an otherwise-profitable wide-N tiling unreachable without dropping the K tile to 32, which costs measurable throughput on its own.

Suggested priority order: (1) graceful K-tiling fallback — small, removes three of the five failures above; (2) a diagnostic that names L0c and the N cap at the deferral point; (3) full `matmul_acc` M/N tiling.

### PR DESCRIPTION
**Pull request by tonibohnlein**
## Summary

Fixes #2232 by adding real M/N tiling for the canonical loop-carried split-K `matmul`/`matmul_acc` form. The rewrite moves the output grid outside the K reduction, so each legal L0C sub-tile completes every source K block and is stored before the next sub-tile begins; the oversized full accumulator is never materialized.

The change also:

- uses one backend-aware physical L0C footprint contract in the chooser and allocators, including padded INT32 accumulator rows;
- pads boundary Mat/L0 operand boxes while preserving their logical `valid_shape`, then propagates that physical/valid distinction through `matmul`, `matmul_acc`, and stores;
- represents PTO compact mode in `TileView` and automatically marks partial Left/Right `tile.extract` results as compact, so TEXTRACT copies only valid INT8 tail data;
- coalesces padded `tile.set_validshape` accumulator initializers with the in-place reduction buffer and rejects irreconcilable distinct Acc-to-Acc copies before codegen; and
- retains `PH-AT-006` for non-canonical, caller-owned oversized `matmul_acc` forms that still cannot be safely retiled.

## Validation

- `cmake --build build --parallel 2`
- Focused host integration suite: **1,470 passed** after rebasing onto current `main` (one untouched current-main parser assertion that expects `NameError` but receives `ParserTypeError` on local Python 3.14 was excluded)
- Issue-specific host coverage: physical L0C accounting, full M/N grid reconstruction, nested K retiling, symbolic and padded boundaries, allocation/reuse, compact-mode serialization, and PTO codegen
- Ascend 910B2 validation before the final rebase: **57/57 AutoTile system-test nodes passed**, including exact INT32 results for `[16, 1152]` and the padded M/N boundary case `[272, 144]` under both PyPTO and PTOAS memory planners
- The formerly failing secondary-K boundary row passed repeatedly under both planners and emitted physical Right/Acc N tails of 32 with logical valid N of 16, with no Acc-to-Acc move or full `[32, 1152]` accumulator

The device run used the feature-equivalent pre-rebase tip `d2526fc6`; the final rebase preserved the feature patches, integrated the new singular `valid_shape` API and fractal-byte documentation from `main`, and was rebuilt and host-revalidated at the PR head.

### CODE REVIEW

File: src/ir/transforms/memory_reuse_pass.cpp [hunk 1]
```diff
@@ -597,7 +599,27 @@ class TopDownRetargeter {
...
+      auto input_tile = GetTileTypeWithMemRef(input_var->GetType());
+      auto output_tile = GetTileTypeWithMemRef(var->GetType());
+      if (!input_tile || !output_tile ||
+          !MemRef::SameAllocation(GetDefinedMemRef(input_tile), GetDefinedMemRef(output_tile))) {
+        return false;
+      }
+      if (!TryRetargetVar(input_var, target, target_memory)) return false;
+      PlanRewrite(var, target, target_memory);
+      return true;
+    }
```
[the remark this answers is not in the document]
Comment (tonibohnlein) [reply src/ir/transforms/memory_reuse_pass.cpp:622]: Added in `97f622b9`. The focused MemoryReuse regression uses a physical 32x32 Acc narrowed to valid 16x16 and verifies that the initializer, `set_validshape` view, accumulator, and loop result share one allocation while preserving the narrowed valid shape.

File: python/pypto/ir/builder.py [hunk 2]
```diff
@@ -689,6 +689,7 @@ def tile_view(
         slayout: ir.TileLayout = ir.TileLayout.none_box,
         fractal: int = 512,
         pad: ir.PadValue = ir.PadValue.null,
+        compact: ir.CompactMode = ir.CompactMode.null,
         span: ir.Span | None = None,
```
[the remark this answers is not in the document]
Comment (tonibohnlein) [reply python/pypto/ir/builder.py]: Fixed in `97f622b9`. `span` is restored to its previous positional slot and `compact` follows it; a regression test exercises an eighth positional `Span` together with keyword `compact`.

File: src/backend/common/pto_ops_elementwise.cpp [hunk 3]
```diff
@@ -567,9 +567,9 @@ void RegisterElementwiseOps(Backend& backend, const std::unordered_set<std::stri
           if (src_offset && dst_offset && src_offset->value_ == dst_offset->value_) {
             const auto src_view = ir::tile_view_semantics::GetEffectiveTileView(*src_tile);
             const auto dst_view = ir::tile_view_semantics::GetEffectiveTileView(*dst_tile);
-            const bool same_layout = src_view.blayout == dst_view.blayout &&
-                                     src_view.slayout == dst_view.slayout &&
-                                     src_view.fractal == dst_view.fractal;
+            const bool same_layout =
+                src_view.blayout == dst_view.blayout && src_view.slayout == dst_view.slayout &&
+                src_view.fractal == dst_view.fractal && src_view.compact == dst_view.compact;
```
[the remark this answers is not in the document]
Comment (tonibohnlein) [reply src/backend/common/pto_ops_elementwise.cpp]: Fixed in `97f622b9`. Same-address move elision now also requires equal `pad` modes, and the new regression proves that differing pad representations retain `pto.tmov`.

**Review by Hzfengsy — APPROVED** [review]
